sudoku_board.cc

```diff
@@ -1,123 +1,181 @@
+      m_possible_options(
+          size, std::vector(size, std::set<int>(POSSIBLE_CELL_VALUES))) {}
+
+SudokuBoard::SudokuBoard(std::vector<std::vector<int>> board_values)
+    : m_size(board_values.size()), m_board_values(std::move(board_values)),
+      m_possible_options(
+          m_size, std::vector(m_size, std::set<int>(POSSIBLE_CELL_VALUES))) {
+  update_possible_options();
 }
 
-SudokuBoard::SudokuBoard(int size, std::vector<std::vector<int>> board_values){
-    m_size = size;
-    m_board_values = board_values; //question: when to use a copy constructor?
-    m_possible_options = std::vector(size, std::vector(size, std::set<int>(POSSIBLE_CELL_VALUES)));
-    update_possible_options();
+// getters
+
+size_t SudokuBoard::get_size() const { return m_size; }
+
+const std::vector<std::vector<int>> &SudokuBoard::get_board_values() const {
+  return m_board_values;
 }
 
-int SudokuBoard::get_size(){
-    return m_size;
+const std::vector<std::vector<std::set<int>>> &
+SudokuBoard::get_possible_options() const {
+  return m_possible_options;
 }
 
-// question: do I need to return a copy and not the original one?
-std::vector<std::vector<int>> SudokuBoard::get_board_values(){
-    return m_board_values;
+const std::set<int> &SudokuBoard::get_cell_options(size_t r, size_t c) const {
+  return m_possible_options[r][c];
 }
 
-std::vector<std::vector<std::set<int>>> SudokuBoard::get_possible_options(){
-    return m_possible_options;
+size_t SudokuBoard::get_missing_cells_amount() const {
+  size_t missing_cells_amount = 0;
+  for (size_t i = 0; i < m_size; i++) {
+    missing_cells_amount += count(m_board_values[i].begin(),
+                                  m_board_values[i].end(), EMPTY_CELL_SIGN);
+  }
+  return missing_cells_amount;
 }
 
-STATUS_OPTION SudokuBoard::get_status(){
-    bool is_win = true;
-    for (int i = 0; i < m_size; i++){
-        for (int j = 0; j < m_size; j++){
-            STATUS_OPTION cell_status = get_cell_status(i, j);
-            if (cell_status == LOSE){
-                return LOSE;
-            }
-            else if (cell_status == PROG){
-                is_win = false;
-            }
-        }
-    }
-    if (is_win){
-        return WIN;
-    }
-    return PROG;
+// setters
+
+void SudokuBoard::set_cell_options(size_t r, size_t c,
+                                   std::set<int> cell_options) {
+  m_possible_options[r][c] = cell_options;
 }
 
-STATUS_OPTION SudokuBoard::get_cell_status(int r, int c){
-    if (get_cell_options(r, c)==std::set<int>({})){
-        if (get_board_values()[r][c]==EMPTY_CELL_SIGN){
-            return LOSE;
-        }
-        else {
-            return WIN;
-        }
-    }
-    return PROG;
+void SudokuBoard::set_cell_value(size_t r, size_t c, int val) {
+  if (get_cell_options(r, c).count(val) <= 0) {
+    throw std::invalid_argument(OPTION_INVALID_ERR_MSG);
+  }
+  m_board_values[r][c] = val;
+  update_possible_options();
```

Comment:
This is more of an algorithmic point of view - do you really need to update the WHOLE board after modifying only a single cell?